File: packages/ik_rules/lib/src/cosmetics/appearance.dart
import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../js_compat.dart';
import '../save/generated/save_models.dart';

enum AppearanceCategory {
  skinTone('skinTone', 'skin_tone', 'Skin tone'),
  hairstyle('hairstyle', 'hairstyle', 'Hairstyle'),
  hairColor('hairColor', 'hair_color', 'Hair color'),
  expression('expression', 'expression', 'Expression'),
  beard('beard', 'beard', 'Beard'),
  genderPresentation('genderPresentation', 'gender_presentation', 'Gender presentation');

  const AppearanceCategory(this.key, this.dataKey, this.label);

  /// Save-schema camelCase key.
  final String key;

  /// Category value in the AppearanceOptions table.
  final String dataKey;

  final String label;
}

/// Resolves a save-schema key (as listed in `appearanceCategories`) to its category.
AppearanceCategory? appearanceCategoryByKey(String key) {
  return AppearanceCategory.values.firstWhereOrNull((category) => category.key == key);
}

String appearanceCategoryLabel(AppearanceCategory category) => category.label;

/// Options for one category, in the table's sort order.
List<AppearanceOptionRow> appearanceOptions(GameDatabase db, AppearanceCategory category) {
  final rows = db.appearanceOptions
      .where((row) => row.raw['Category'] == category.dataKey)
      .toList();
  mergeSort(
    rows,
    compare: (a, b) => jsCompareThen(
      jsNumberOrZero(a.raw['Sort Order']) - jsNumberOrZero(b.raw['Sort Order']),
      () => 0,
    ),
  );
  return rows;
}

AppearanceOptionRow? appearanceOptionById(GameDatabase db, String optionId) {
  return db.appearanceOptions.firstWhereOrNull(
    (row) => row.raw['Appearance Option ID'] == optionId,
  );
}

bool isValidAppearanceOption(GameDatabase db, AppearanceCategory category, String optionId) {
  return appearanceOptions(db, category).any((row) => row.raw['Appearance Option ID'] == optionId);
}

/// The first option per category by sort order, falling back to the baselines.
PlayerAppearance defaultAppearance(GameDatabase db) {
  String pick(AppearanceCategory category, String fallback) {
    final first = appearanceOptions(db, category).firstOrNull;
    final optionId = first?.raw['Appearance Option ID'];
    return optionId is String ? optionId : fallback;
  }

  return PlayerAppearance(
    skinTone: pick(AppearanceCategory.skinTone, defaultSkinToneId),
    hairstyle: pick(AppearanceCategory.hairstyle, defaultHairstyleId),
    hairColor: pick(AppearanceCategory.hairColor, defaultHairColorId),
    expression: pick(AppearanceCategory.expression, defaultExpressionId),
    beard: pick(AppearanceCategory.beard, defaultBeardId),
    genderPresentation: pick(AppearanceCategory.genderPresentation, defaultGenderPresentationId),
  );
}

String appearanceSelection(PlayerAppearance appearance, AppearanceCategory category) {
  return switch (category) {
    AppearanceCategory.skinTone => appearance.skinTone,
    AppearanceCategory.hairstyle => appearance.hairstyle,
    AppearanceCategory.hairColor => appearance.hairColor,
    AppearanceCategory.expression => appearance.expression,
    AppearanceCategory.beard => appearance.beard,
    AppearanceCategory.genderPresentation => appearance.genderPresentation,
  };
}

/// Re-selectable freely: appearance carries no gameplay effect. Null when the
/// option does not belong to the category.
PlayerSave? setAppearanceOption(
  GameDatabase db,
  PlayerSave save,
  AppearanceCategory category,
  String optionId,
) {
  if (!isValidAppearanceOption(db, category, optionId)) return null;
  final appearance = switch (category) {
    AppearanceCategory.skinTone => save.appearance.copyWith(skinTone: optionId),
    AppearanceCategory.hairstyle => save.appearance.copyWith(hairstyle: optionId),
    AppearanceCategory.hairColor => save.appearance.copyWith(hairColor: optionId),
    AppearanceCategory.expression => save.appearance.copyWith(expression: optionId),
    AppearanceCategory.beard => save.appearance.copyWith(beard: optionId),
    AppearanceCategory.genderPresentation => save.appearance.copyWith(genderPresentation: optionId),
  };
  return save.copyWith(appearance: appearance);
}

File: src/game/cosmetics/appearance.ts
import type { AppearanceCategoryKey, AppearanceOptionRow, GameDatabase } from '../data/types'
import {
  APPEARANCE_CATEGORIES,
  DEFAULT_BEARD_ID,
  DEFAULT_EXPRESSION_ID,
  DEFAULT_GENDER_PRESENTATION_ID,
  DEFAULT_HAIRSTYLE_ID,
  DEFAULT_HAIR_COLOR_ID,
  DEFAULT_SKIN_TONE_ID,
  type AppearanceCategory,
  type PlayerAppearance,
  type PlayerSave,
} from '../save/types'

export { APPEARANCE_CATEGORIES }
export type { AppearanceCategory }

/** Save-schema camelCase category -> data-table snake_case Category value. */
const CATEGORY_DATA_KEY: Record<AppearanceCategory, AppearanceCategoryKey> = {
  skinTone: 'skin_tone',
  hairstyle: 'hairstyle',
  hairColor: 'hair_color',
  expression: 'expression',
  beard: 'beard',
  genderPresentation: 'gender_presentation',
}

const CATEGORY_LABELS: Record<AppearanceCategory, string> = {
  skinTone: 'Skin tone',
  hairstyle: 'Hairstyle',
  hairColor: 'Hair color',
  expression: 'Expression',
  beard: 'Beard',
  genderPresentation: 'Gender presentation',
}

export function appearanceCategoryLabel(category: AppearanceCategory): string {
  return CATEGORY_LABELS[category]
}

/** Options for one Appearance category, sorted for stable display order. */
export function appearanceOptions(db: GameDatabase, category: AppearanceCategory): AppearanceOptionRow[] {
  const key = CATEGORY_DATA_KEY[category]
  return db.AppearanceOptions.filter((row) => row.Category === key).sort(
    (a, b) => (a['Sort Order'] ?? 0) - (b['Sort Order'] ?? 0),
  )
}

export function appearanceOptionById(
  db: GameDatabase,
  optionId: string,
): AppearanceOptionRow | undefined {
  return db.AppearanceOptions.find((row) => row['Appearance Option ID'] === optionId)
}

export function isValidAppearanceOption(
  db: GameDatabase,
  category: AppearanceCategory,
  optionId: string,
): boolean {
  return appearanceOptions(db, category).some((row) => row['Appearance Option ID'] === optionId)
}

const FALLBACK_DEFAULTS: PlayerAppearance = {
  skinTone: DEFAULT_SKIN_TONE_ID,
  hairstyle: DEFAULT_HAIRSTYLE_ID,
  hairColor: DEFAULT_HAIR_COLOR_ID,
  expression: DEFAULT_EXPRESSION_ID,
  beard: DEFAULT_BEARD_ID,
  genderPresentation: DEFAULT_GENDER_PRESENTATION_ID,
}

/** First (by Sort Order) option per category, falling back to the baseline constants. */
export function defaultAppearance(db: GameDatabase): PlayerAppearance {
  const result = { ...FALLBACK_DEFAULTS }
  for (const category of APPEARANCE_CATEGORIES) {
    const first = appearanceOptions(db, category)[0]
    if (first) result[category] = first['Appearance Option ID']
  }
  return result
}

/**
 * One category swapped out, without a save to put it in.
 *
 * Unvalidated on purpose: character creation drives this from the sliders,
 * which only offer options the tables list.
 */
export function withAppearanceOption(
  appearance: PlayerAppearance,
  category: AppearanceCategory,
  optionId: string,
): PlayerAppearance {
  return { ...appearance, [category]: optionId }
}

/** Freely re-selectable at any time (skin tone, hairstyle, etc. carry no gameplay effect). */
export function setAppearanceOption(
  db: GameDatabase,
  save: PlayerSave,
  category: AppearanceCategory,
  optionId: string,
): PlayerSave | null {
  if (!isValidAppearanceOption(db, category, optionId)) return null
  return { ...save, appearance: withAppearanceOption(save.appearance, category, optionId) }
}
